<?xml version="1.0"?>
<launch>
  <arg name="arm_1_joint_cmd_pos_name" default="/ur5e/arm_1_joint/cmd_pos" />
  <arg name="arm_2_joint_cmd_pos_name" default="/ur5e/arm_2_joint/cmd_pos" />
  <arg name="arm_3_joint_cmd_pos_name" default="/ur5e/arm_3_joint/cmd_pos" />
  <arg name="up_down_lower_limit" default="[-2.8]" />
  <arg name="up_down_upper_limit" default="[2.8]" />
  <arg name="left_right_lower_limit" default="[-2.8]" />
  <arg name="left_right_upper_limit" default="[2.8]" />
  <arg name="near_far_lower_limit" default="[-2.8, -2.8]" />
  <arg name="near_far_upper_limit" default="[2.8, 2.8]" />
  <arg name="joint_states_topic" default="/ur5e/joint_states" />
  <arg name="shoulder_frame" default="base_link" />
  <arg name="tcp_frame" default="ee_link" />
  <arg name="joint_names" default="['shoulder_pan_joint', 'shoulder_lift_joint', 'elbow_joint', 'wrist_1_joint', 'wrist_2_joint', 'wrist_3_joint']" />
  <arg name="arm_3_joint_index" default="3" />
  <arg name="follow_joint_trajectory_param" default="/ur5e/arm_controller/follow_joint_trajectory" />
  <arg name="next_pos_delta_factor" default="0.7" />
  <arg name="map_phi_to_2pi" default="False" />
  <arg name="pred_pos_frame" default="base_link" />
  <arg name="pred_pos_topic" default="/pred_pos_3" />
  <arg name="target_position_topic" default="/ur5e/target_position" />
  <arg name="error_topic" default="/ur5e/error" />
  <arg name="sphere_marker_pub_topic" default="/ur5e/sphere_marker" />
  <arg name="base_network_class_data_pub_topic" default="/ur5e/base_network_class_data_pub_topic" />
  <arg name="error_class_data_topic" default="/ur5e/error_class_data_topic" />
  <arg name="feedback_data_pub_topic" default="/ur5e/feedback_data_pub_topic" />

  <node name="pub_ur5e_target_as_marker" pkg="target_reaching_nengo" type="pub_target_pos_as_vis_marker.py" output="screen">
    <param name="target_position_topic" value="$(arg target_position_topic)" />
    <param name="sphere_marker_pub_topic" value="$(arg sphere_marker_pub_topic)" />
  </node>

  <node name="pub_ur5e_pred_pos_as_target" pkg="target_reaching_nengo" type="pub_ur5e_target_position.py" output="screen">
    <param name="pred_pos_frame" value="$(arg pred_pos_frame)" />
    <param name="pred_pos_topic" value="$(arg pred_pos_topic)" />
    <param name="target_position_topic" value="$(arg target_position_topic)" />
    <param name="error_topic" value="$(arg error_topic)" />
  </node>

  <node name="Main_TR_CL_nengo_node_ur5e" pkg="target_reaching_nengo" type="main_TR_CL_node.py" output="screen">
    <param name="arm_1_joint_cmd_pos_name" value="$(arg arm_1_joint_cmd_pos_name)" />
    <param name="arm_2_joint_cmd_pos_name" value="$(arg arm_2_joint_cmd_pos_name)" />
    <param name="arm_3_joint_cmd_pos_name" value="$(arg arm_3_joint_cmd_pos_name)" />
    <rosparam param="up_down_lower_limit" subst_value="True">$(arg up_down_lower_limit)</rosparam>
    <rosparam param="up_down_upper_limit" subst_value="True">$(arg up_down_upper_limit)</rosparam>
    <rosparam param="left_right_lower_limit" subst_value="True">$(arg left_right_lower_limit)</rosparam>
    <rosparam param="left_right_upper_limit" subst_value="True">$(arg left_right_upper_limit)</rosparam>
    <rosparam param="near_far_lower_limit" subst_value="True">$(arg near_far_lower_limit)</rosparam>
    <rosparam param="near_far_upper_limit" subst_value="True">$(arg near_far_upper_limit)</rosparam>
    <param name="joint_states_topic" value="$(arg joint_states_topic)" />
    <param name="shoulder_frame" value="$(arg shoulder_frame)" />
    <param name="tcp_frame" value="$(arg tcp_frame)" />
    <rosparam param="joint_names" subst_value="True">$(arg joint_names)</rosparam>
    <param name="arm_3_joint_index" value="$(arg arm_3_joint_index)" />
    <param name="next_pos_delta_factor" value="$(arg next_pos_delta_factor)" />
    <param name="map_phi_to_2pi" value="$(arg map_phi_to_2pi)" />
    <param name="base_network_class_data_pub_topic" value="$(arg base_network_class_data_pub_topic)" />
    <param name="error_class_data_topic" value="$(arg error_class_data_topic)" />
    <param name="feedback_data_pub_topic" value="$(arg feedback_data_pub_topic)" />
    <param name="target_position_topic" value="$(arg target_position_topic)" />
    <param name="error_topic" value="$(arg error_topic)" />
  </node>

  <node name="ur5e_mapping" pkg="target_reaching_nengo" type="target_reaching_to_ur5e_mapping.py" output="screen">
    <param name="arm_1_joint_cmd_pos_name" value="$(arg arm_1_joint_cmd_pos_name)" />
    <param name="arm_2_joint_cmd_pos_name" value="$(arg arm_2_joint_cmd_pos_name)" />
    <param name="arm_3_joint_cmd_pos_name" value="$(arg arm_3_joint_cmd_pos_name)" />
    <param name="joint_states_topic" value="$(arg joint_states_topic)" />
    <rosparam param="joint_names" subst_value="True">$(arg joint_names)</rosparam>
    <param name="arm_3_joint_index" value="$(arg arm_3_joint_index)" />
    <param name="follow_joint_trajectory_param" value="$(arg follow_joint_trajectory_param)" />
  </node>
</launch>
